soc: mimxrt595: Do not dynamically change pins in soc power driver

mmahadevan108 opened the pull request:
Pin configuration is board specific and should be moved out of the soc folder to the board level.
Use pm_notifier feature to achieve this.

gmarull reviewed (changes requested):
please describe bits in DT, this looks like a hack in a board c file

dleach02 commented:
> please describe bits in DT, this looks like a hack in a board c file

@gmarull, not sure what your request is. This is moving code from the soc file to the board file because having it in the SOC file doesn't scale for custom boards.

dleach02 commented on boards/arm/mimxrt595_evk/board.c:
```diff
@@ -95,6 +96,54 @@ static int board_config_pmic(void)
+
+static __ramfunc void set_deepsleep_pin_config(void)
+{
+	/* Backup Pin configuration. */
+	isp_pin[0] = IOPCTL->PIO[1][15];
+	isp_pin[1] = IOPCTL->PIO[3][28];
+	isp_pin[2] = IOPCTL->PIO[3][29];
+
+	/* Disable ISP Pin pull-ups and input buffers to avoid current leakage */
+	IOPCTL->PIO[1][15] = 0;
+	IOPCTL->PIO[3][28] = 0;
+	IOPCTL->PIO[3][29] = 0;
+}
+
+static __ramfunc void restore_deepsleep_pin_config(void)
+{
+	/* Restore the Pin configuration. */
+	IOPCTL->PIO[1][15] = isp_pin[0];
+	IOPCTL->PIO[3][28] = isp_pin[1];
+	IOPCTL->PIO[3][29] = isp_pin[2];
```
Maybe. But as pointed out, this is existing code in the upstream that is being moved from one file to another.

gmarull replied:
that's not a reason to not do things properly, sorry.

mmahadevan108 committed:
soc: nxp: rt5xx: Remove deepsleep pin changes

SOC level code should not be dynamically changing pin configurations.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
@gmarull, I have updated to use `DTS` and the `pinctrl` api.

gmarull commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -127,6 +127,12 @@
 		regulator-boot-on;
 	};
+
+	isp: isp {
```
even if not used, doesn't this have an address and belong to `soc/` node?

mmahadevan108 replied:
no it is not have an address.

gmarull replied:
so then, what is this node describing exactly? If you really need to manipulate some random GPIOs that turn out to have a predefined functionality (but not connected to any peripheral, like say I2C), why not use the GPIO API, just storing them as `*-gpios` in DT?

mmahadevan108 replied:
@gmarull, any additional comments your end.

decsny reviewed (changes requested):
I was approving of the original edition of the PR, I dont think pinctrl makes sense here, because it's barely even a pinctrl state we are describing but one single bit that needs to be temporarily flipped during PM. the actual totality of the state that the pins are meant to have before and after this transition could be theoretically anything

It could be debated if using GPIO API is doable or correct. But doing so would require changes to the GPIO driver which is outside the scope of the PR, and would introduce unnecessary latency and overhead to entering and exiting power modes. Also, this isn't even a place where we need to use a generic API in my opinion because it is something specific to this board, so the code doesn't need to be portable to other boards, and it's also not something I would expect to have any value from being in DT. @mmahadevan108 correct me if I'm wrong, if you expect this to have value being in DT for some reason

dleach02 commented:
@gmarull, @mmahadevan108, @decsny, please join this week's dev review meeting so we can get this PR merged.

decsny commented:
the current implementation seems to have conceded that this does not need to happen at every entrance/exit of deepsleep but only once in init to ensure those bits are cleared, but we are doing it through the pinctrl api which as far as I can tell has no guarantee of doing that, and just is happening by luck of the implementation

gmarull commented:
I won't be able to join dev-review. However, I leave here my questions/requests:

- Configuration needs to happen in Devicetree. This doesn't necessarily mean usage of pinctrl or GPIO, just provide enough input to provide feedback. What is the code exactly doing, and for what reason?
- Board level code needs a good justification. Is the action specific to the board, ie, depending on the board layout? Isn't this something other boards based on the same SoC will need? Because then, this should be handled by a driver or soc code, no custom board code.

mmahadevan108 commented:
> I won't be able to join dev-review. However, I leave here my questions/requests:
> 
> * Configuration needs to happen in Devicetree. This doesn't necessarily mean usage of pinctrl or GPIO, just provide enough input to provide feedback. What is the code exactly doing, and for what reason?
> * Board level code needs a good justification. Is the action specific to the board, ie, depending on the board layout? Isn't this something other boards based on the same SoC will need? Because then, this should be handled by a driver or soc code, no custom board code.

This code is indeed board specific and hence moved out of `soc.c`

mmahadevan108 commented:
@gmarull @decsny @DerekSnell, I have updated this implementation to keep the generic portion inside `soc.c`. This code uses the information provided from the board level `pincrtl` file to configure the boot ISP pins.
@MaureenHelm I have created a generic binding for the ISP pins, thank you for this suggestion. This should usable by other NXP SoC's

thedjnK commented on soc/arm/nxp_imx/rt5xx/soc.c:
```diff
@@ -512,6 +513,20 @@ void __weak imxrt_deinit_display_interface(void)
 
 #endif
 
+static int nxp_rt500_late_init(void)
+{
+	/* Declare pin configuration state for isp pin here */
+	PINCTRL_DT_DEFINE(DT_NODELABEL(isp));
+
+	/* Apply pinctrl state directly, since there is no isp device driver */
+	pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(DT_NODELABEL(isp)),
+			    PINCTRL_STATE_DEFAULT);
+
+	return 0;
+}
+
+SYS_INIT(nxp_rt500_late_init, PRE_KERNEL_2, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
Seems like you would want to check the status or this will cause a build failure on any board using this soc which does not define or enable the ISP pins?

mmahadevan108 replied:
This should not have an impact on boards that do not define the pinctrl setting for their board.

mmahadevan108 replied:
@thedjnK, I have added a build time check to avoid these extra functions from being called on platforms that do not require reconfiguring the boot pins.

DerekSnell reviewed (dismissed):
thank you @mmahadevan108 , LGTM.

gmarull commented on dts/bindings/arm/nxp,bootpin.yaml:
```diff
@@ -0,0 +1,17 @@
+
+compatible: "nxp,bootpin"
+
+include: [base.yaml, pinctrl-device.yaml]
```
base is not required

gmarull commented on dts/bindings/arm/nxp,bootpin.yaml:
```diff
@@ -0,0 +1,17 @@
+  declares pinctrl properties, so that the user can configure boot pins
+  at the board level.
+
+compatible: "nxp,bootpin"
```
this looks like an incorrect usage of compatible because this is not an actual device/IP. Maybe `nxp-boot-pins`, like Arduino/ST morpho headers are specified.

gmarull commented on dts/bindings/arm/nxp,bootpin.yaml:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  NXP Boot pins binding. This binding does not represent a full driver, but
+  declares pinctrl properties, so that the user can configure boot pins
+  at the board level.
```
Remember that DT describes hardware, so comments about system-level stuff (driver, pinctrl, etc.) should be avoided. Just describe What this binding is for, and maybe a usage example.

gmarull commented on dts/arm/nxp/nxp_rt5xx_common.dtsi:
```diff
@@ -679,6 +679,11 @@
 		};
 	};
+
+	isp: isp {
+		compatible = "nxp,bootpin";
```
I find naming confusing: isp, bootpin, align or clarify please.

gmarull commented on soc/arm/nxp_imx/rt5xx/soc.c:
```diff
@@ -512,6 +513,29 @@ void __weak imxrt_deinit_display_interface(void)
 
 #endif
 
+static int nxp_rt500_late_init(void)
+{
+#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(isp), nxp_bootpin, okay)
+
+	int err;
+
+	/* Declare pin configuration state for isp pin here */
+	PINCTRL_DT_DEFINE(DT_NODELABEL(isp));
+
+	/* Apply pinctrl state directly, since there is no isp device driver */
+	err = pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(DT_NODELABEL(isp)),
+				  PINCTRL_STATE_DEFAULT);
+
+	if (err) {
+		return err;
+	}
+
+#endif
+	return 0;
+}
+
+SYS_INIT(nxp_rt500_late_init, PRE_KERNEL_2, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
please guard the whole function/SYS_INIT, there's no point in running an empty SYS_INIT.

gmarull commented on boards/arm/mimxrt595_evk/board.c:
```diff
@@ -308,7 +308,6 @@ static int mimxrt595_evk_init(void)
 
 #endif
 
-
```
?

gmarull reviewed (changes requested):
some nitpicks

decsny reviewed (changes requested):
at this point the changes are so ridiculous, and I would prefer now to just call a gpio configure function in the board init, which makes the most architectural sense, since what we really are trying to do is configure a gpio due to a hardware configuration of the board 

in fact, @danieldegrasse has pointed out in offline discussion that a paradigm in DT already exists exactly for this situation, gpio-hog

this will require a change to the gpio driver but it will be probably the best solution here to resolve these blocks, I think a gpio hog makes sense now that we decided this is an init problem

decsny commented on soc/arm/nxp_imx/rt5xx/soc.c:
```diff
@@ -512,6 +513,29 @@ void __weak imxrt_deinit_display_interface(void)
+	return 0;
+}
+
+SYS_INIT(nxp_rt500_late_init, PRE_KERNEL_2, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
lets not put any pinctrl calls in the soc file, please

nordicjm commented:
> at this point the changes are so ridiculous

I cannot agree with you here about this and think the suggestions @gmarull has made are sensible, I'm sorry you don't see it the same way.

decsny commented:
> > at this point the changes are so ridiculous
> 
> I cannot agree with you here about this and think the suggestions @gmarull has made are sensible, I'm sorry you don't see it the same way.

the reason I say it's ridiculous is because the PR was originally just moving code, and all the nxp platform maintainer agreed on this, and this PR touches nothing except nxp soc/boards.... but now we have all this multilevel init code and new pinctrl binding and dts entries for something as simple as two gpios that just need to be initialized to a certain initial configuration, it's too much.

gpio hog is the exact use of this, so we can use that, I don't disagree with @gmarull that configuration should go in DT, I just think it shouldn't look like this, if we have to do it somehow then it should be as simple as the situation calls for, ie gpio hog, not a whole binding and soc init level created for a momentary state of two gpio pins

@gmarull do you disagree with the idea of representing this in DT as gpio hog node? and removing all this pinctrl stuff

gmarull commented:
> > > at this point the changes are so ridiculous
> > 
> > 
> > I cannot agree with you here about this and think the suggestions @gmarull has made are sensible, I'm sorry you don't see it the same way.
> 
> the reason I say it's ridiculous is because the PR was originally just moving code, and all the nxp platform maintainer agreed on this, and this PR touches nothing except nxp soc/boards.... but now we have all this multilevel init code and new pinctrl binding and dts entries for something as simple as two gpios that just need to be initialized to a certain initial configuration, it's too much.
> 

soc/ is plagued with tons of vendor crap, but when there’s an opportunity to improve I think there’s no excuse to not do it. I think my request has been fairly reasonable: keep description in DT. How it’s then used can be discussed.

> gpio hog is the exact use of this, so we can use that, I don't disagree with @gmarull that configuration should go in DT, I just think it shouldn't look like this, if we have to do it somehow then it should be as simple as the situation calls for, ie gpio hog, not a whole binding and soc init level created for a momentary state of two gpio pins
> 
> @gmarull do you disagree with the idea of representing this in DT as gpio hog node? and removing all this pinctrl stuff

GPIO hogs are a good solution if this is a one-shot configuration. I still miss information to say if it’s the best though, and after all iterations I’m quite confused on what you are really trying to achieve here.

decsny commented:
> GPIO hogs are a good solution if this is a one-shot configuration. I still miss information to say if it’s the best though, and after all iterations I’m quite confused on what you are really trying to achieve here.

The situation is this: These 3 pins are simply gpios that are read by the rom at boot to determine the boot sequence. The ROM enables the pins' input buffer in order to read them. So far on any RT500 we tested, the ROM disables the input buffer once it is done consuming them. But we had some concern that there may be some chip revision where the ROM doesn't do this for some reason. And since these pins are almost always connected to pull-up/pull-down resistors to select boot configuration, in case this happens then there would be some power leakage, which we want to avoid in the default configuration in zephyr. I actually doubt there is even any chip revision which doesn't do this, and we probably don't need any code at all, because the default value of the pin should be input buffer disabled, but we decided to put this code just to precaution in case these pins are not used by a user application later. In case the user touches these pins at all then they will be responsible for the configuration at that point. Which is why I orginally just want to write these bits directly in the soc init (all we are doing is ensuring the reset value after ROM boot), but if we have to use devicetree to describe this, then I guess a gpio hog is fine. It still seems slightly pointless to me since any user who is ever configuring this pin will overwrite the gpio hog configuration since all it is describing is the reset value state of a pin.

So yes, what we are trying to accomplish right now is a one-shot configuration of one gpio property of one pin, but it's not a special configuration, it's just a precautionary restoration of reset values right after rom boot in init. And this restoration should happen BEFORE any board or app level configuration would overwrite it by using the pin. So even a gpio hog would just kind of be dead weight in a user board definition if they arent using the pin, but it's fine if we have to use DT, it's the lightest approach, other than just writing to the registers their reset value in soc init

gmarull commented:
> > GPIO hogs are a good solution if this is a one-shot configuration. I still miss information to say if it’s the best though, and after all iterations I’m quite confused on what you are really trying to achieve here.
> 
> The situation is this: These 3 pins are simply gpios that are read by the rom at boot to determine the boot sequence. The ROM enables the pins' input buffer in order to read them. On most chip revisions, ROM disables the input buffer once it is done consuming them, but potentially some chip revisions the ROM doesn't do this. And since these pins are almost always connected to pull-up/pull-down resistors to select boot configuration, this would cause some power leakage, which we want to avoid in the default configuration in zephyr.
> 
> So basically we are trying to make sure we restore the reset value of the pin is all. So yes it is a one time configuration. But it is not a special configuration, any user use of the pin would override this "configuration" as once again, it is just the reset value of the pin. I would most prefer to just write these registers during init as I would expect any user that actually uses these pins for something will have a different node in DT describing what it is supposed to do. The functions of this pin are either gpio, ctimer trigger, or spi.
> 
> What we are trying to do is restore the reset value where it is just basically a high impedance and does nothing, then I would expect either the ctimer, spi, or gpio driver will be used to initialize the pin to it's actual function at a later time. So that is why my preferred solution is to just write into these registers in the SOC init, since this is only dealing with a potential ROM errata, we're not really trying to configure something.
> 
> @gmarull does this info help you to understand the situation better?
> 
> And also I should clarify, I am disagreeing with the original intent of the PR to move this to board level, because I think these ISP-designated GPIOs for the ROM are basically always going to be connected to pull-up pull-down, and will probably always have these power leakage problem if the ROM has this errata. So, I don't think it's something we need to burden every customer/user board definition with, and doesn't really belong in board level because once again, we are just trying to restore reset value of the pins as early as possible so that a user is not surprised by this power leakage if they didn't enable the use of the pin in their configuration. What I think this PR should do is remove the code that changes these pin configurations dynamically at any point after the earliest init, because that was wrong and unnecessary and we got some customer complaint about dynamically touching pin configurations at soc level. But having it in the earliest init sequence as a one time write just to restore the reset value in case of ROM errata seems reasonable to me. But calling it a pinctrl configuration and declaring DT for it just seems innaccurate for this situation in my opinion.
> 
> If we have to use GPIO hog for this then that will require us to rework our GPIO driver. And we are trying to get this fix into 3.6, so I don't think we want to do that. Again, all we are doing is just restoring reset value, not trying to use the pin for anything. If a user wants to use the pin they will have a different DT description for an actual function of the pin.

From what I understand, this is what I'd suggest:

- GPIO hogs are also a good choice, as you can define them only if needed at board level (ie, if your board has a buggy SoC). You just need to make sure that hogs are run before other peripherals use pinctrl (they just run after GPIO driver init, which tipically happens early).
- Hardcoding such stuff in SoC is fine, as it doesn't seem to be configurable at all. However, it deserves a good explanation as you've done now, so people can understand what's going on (ie, there's a ROM, it may be buggy, pins may have pulls connected, we need to avoid leakage, etc.)

In any case, all the details you've written now should be stored somewhere, eg commit message.

Commit:
soc: rt5xx: Restore ISP pins state in soc init

ROM configures the ISP boot pins as gpio to determine what boot mode to
be in. But some ROM revisions have a bug where they do not restore the
reset state of these pins before booting application. This can cause
power leakage on these pins and is not an intended configuration from
Zephyr user/board point of view, so restore the reset state as part of
early SOC init (disable the pins). Configuration of pins should be
left up to app/board devicetree.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@gmarull thank you for the suggestion, I put the info in a comment and the commit message, can you review if you think they are sufficiently documented
